src/Bicep.Core/Analyzers/Linter/Rules/PreferBarePropertyNamesRule.cs

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Bicep.Core.CodeAction;
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Parsing;
+using Bicep.Core.Semantics;
+using Bicep.Core.Syntax;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+namespace Bicep.Core.Analyzers.Linter.Rules
+{
+    public sealed class PreferBarePropertyNamesRule : LinterRuleBase
+    {
+        public new const string Code = "prefer-bare-property-names";
+
+        public PreferBarePropertyNamesRule() : base(
+            code: Code,
+            description: CoreResources.PreferBarePropertyNamesRuleDescription,
+            docUri: new Uri($"https://aka.ms/bicep/linter/{Code}"))
+        { }
+
+        public override IEnumerable<IDiagnostic> AnalyzeInternal(SemanticModel model)
+        {
+            var spanFixes = new Dictionary<TextSpan, CodeFix>();
+            var visitor = new Visitor(spanFixes);
+            visitor.Visit(model.SourceFile.ProgramSyntax);
+
+            return spanFixes.Select(kvp => CreateFixableDiagnosticForSpan(kvp.Key, kvp.Value));
+        }
+
+        private sealed class Visitor : SyntaxVisitor
+        {
+            private static readonly Regex VALID_IDENTIFIER_PATTERN = new(@"^[a-zA-Z_]\w*$");
+            private readonly Dictionary<TextSpan, CodeFix> spanFixes;
+
+            public Visitor(Dictionary<TextSpan, CodeFix> spanFixes)
+            {
+                this.spanFixes = spanFixes;
+            }
+
+            public override void VisitObjectPropertySyntax(ObjectPropertySyntax syntax)
+            {
+                if (TryGetValidIdentifierToken(syntax.Key, out string? literal))
+                {
+                    AddCodeFix(syntax.Key.Span, literal, CoreResources.PreferBarePropertyNamesDeclarationFixTitle);
+                }
+
+                base.VisitObjectPropertySyntax(syntax);
+            }
+
+            public override void VisitArrayAccessSyntax(ArrayAccessSyntax syntax)
+            {
+                if (TryGetValidIdentifierToken(syntax.IndexExpression, out string? literal))
+                {
+                    AddCodeFix(TextSpan.Between(syntax.OpenSquare, syntax.CloseSquare), $".{literal}", CoreResources.PreferBarePropertyNamesDereferenceFixTitle);
+                }
+
+                base.VisitArrayAccessSyntax(syntax);
+            }
+
+            private void AddCodeFix(TextSpan span, string replacement, string description)
+            {
+                var codeReplacement = new CodeReplacement(span, replacement);
+                var fix = new CodeFix(description, true, CodeFixKind.QuickFix, codeReplacement);
```

Comment:
<img width="520" alt="image" src="https://user-images.githubusercontent.com/6913354/167747294-eab4138a-81bd-4a27-a31e-fbd1e9c6fb7a.png">

For code fixes, we like to see more explicit titles on what exactly will be done.  Suggested for this scenario for the fix's title:
```
    Replace array syntax with '.sku'
```

cc @ucheNkadiCode